### ISSUE TITLE
Login / Logout Function

### ISSUE THREAD

**Issue by lucasdiedrich** [lucasdiedrich/node-freeipa#14]
The library should have both functions to authenticate the user and it should not use cookie cache.

**Comment by zyxep** [comment]
Are there only 1 way to auth and that is the option way?
So I can't pass in data from a form?

**Comment by lucasdiedrich** [comment]
Hey @zyxep , how are you?

The idea of this issue was to create another way of authenticating the user, but it was never developed, authentication by options is the only available for now.

Thanks.

**Comment by zyxep** [comment]
I'm good thanks.
Still trying to continue with my self-service page based on NuxtJS & a ExpressJS API.

sounds like it's going to be a pain in the *** for me to do the auth of a user. So the user can change their profile.

**Comment by lucasdiedrich** [comment]
Yep, that's why i created this issue, when i was going to authenticate the
user i realized the we need this function, but as noone requested that i
didn't developed it.

I can plan to make develop this issue next week. What you think?

NuxtJS seems really nice to me, excellent choice.


Em sex, 15 de mar de 2019 às 09:28, zyxep <notifications@github.com>
escreveu:

**Comment by zyxep** [comment]
That would be amazing, this is self-service page is an ongoing project for me at work :)
Since I don't like how FreeIPA do it.

So I managed right now to build a reset with Nuxt,Express & Redis :)
Though i did have to add my own user to their password policy so my account can change user password without it expire.

**Comment by lucasdiedrich** [comment]
YEAP, this is how Freeipa works,  when a third user changes someones password, it marks as expired, forcing the user itself to change the password, this is a security measurement. 

Ok going to make the plans to make this at next week @zyxep  ;)

**Comment by zyxep** [comment]
Hey :D

What is latest? I can see you also have https://github.com/lucasdiedrich/passport-freeipa which I thought using.
Otherwise would I use http://www.passportjs.org/packages/passport-ldapauth/

**Comment by lucasdiedrich** [comment]
Hey @zyxep, unfortunetly i've to ask you to wait another week, i'm out of
time, but promise to take a look at this weekend, sorry for that.

Thanks.

Em qua, 27 de mar de 2019 às 06:32, zyxep <notifications@github.com>
escreveu:

**Comment by zyxep** [comment]
it's okay :)
I managed to get a login to work with passport-LDAP
Doing this:

```JavaScript
app.post('/login', passport.authenticate('ldapauth', {session: false}), function(req, res) {
  const user = req.user
  res.send({
    user: {
      id: user.uidNumber,
      dn: user.dn,
      givenName: user.givenName,
      lastName: user.sn,
      uid: user.uid,
      cn: user.cn,
      displayName: user.displayName,
      initials: user.initials,
      mail: user.mail
    },
    token: jwt.sign({id: user.uidNumber, uid: user.uid}, process.env.PRIVATEKEY)
  })
})

app.get('/me', function(req, res) {
  const unverifiedToken = req.headers.authorization.split('Bearer ')[1]
  try {
    const decoded = jwt.verify(unverifiedToken, process.env.PRIVATEKEY)

    ipa.user_show([decoded.uid], { all: true }).then((result) => {
      if (result.error) {
        res.status(403).send({ error: 'No user found' })
      } else {
        const user = result
        res.send({
          user: {
            id: user.uidnumber[0],
            dn: user.dn,
            givenName: user.givenname[0],
            lastName: user.sn[0],
            uid: user.uid[0],
            cn: user.cn[0],
            displayName: user.displayname[0],
            initials: user.initials[0],
            mail: user.mail[0],
            groups: user.memberof_group

          }
        })
      }
    })
  } catch(error) {
    res.status(403).send(error)
  }
})
```

**Comment by lucasdiedrich** [comment]
Closing this for now!
